### PR TITLE
Added X-Fi check and convert surround FLOAT to SHORT when it is detected

### PR DESCRIPTION
**Pull request by LAGonauta**
Follow up from https://bugs.dolphin-emu.org/issues/8711

The OpenAL backend wasn't working on hardware based OpenAL cards (basically only the X-Fi), stereo mode generated buffer queue errors and never played anything, and surround mode was very loud with lots of clipping.
This happens because the X-Fi doesn't support float point operation, only fixed point (even though the drivers reports that it supports).
I added a check to detect a X-Fi, and when it does it converts float to short.
Stereo is perfect now (I used the code that was already there), but for some reason on surround there is some clipping (and this clipping happens even if I force short mode with OpenAL Soft).
I couldn't find why this happens, sadly, but now it is usable.

### CODE REVIEW

File: Source/Core/AudioCommon/OpenALStream.cpp [hunk 1]
```diff
@@ -146,18 +146,39 @@ void OpenALStream::SoundLoop()
 	memset(uiBuffers, 0, numBuffers * sizeof(ALuint));
 	uiSource = 0;
 
+	// Checks if a X-Fi is being used. If it is, disable FLOAT32 support.
```
Comment (JosJuice) [inline Source/Core/AudioCommon/OpenALStream.cpp]: This comment should include why this is done, not just that it's done.

File: Source/Core/AudioCommon/OpenALStream.cpp [hunk 2]
```diff
@@ -146,18 +146,39 @@ void OpenALStream::SoundLoop()
...
+	if (strstr(alGetString(AL_RENDERER), "X-Fi") != NULL)
```
Comment (JosJuice) [inline Source/Core/AudioCommon/OpenALStream.cpp]: We prefer `nullptr` over `NULL` (or as another option, I think you can leave out the ` != NULL` part entirely).

File: Source/Core/AudioCommon/OpenALStream.cpp [hunk 3]
```diff
@@ -146,18 +146,39 @@ void OpenALStream::SoundLoop()
...
 	// Short Silence
-	memset(sampleBuffer, 0, OAL_MAX_SAMPLES * numBuffers * FRAME_SURROUND_FLOAT);
+	if (float32_capable)
+	{
+		memset(sampleBuffer, 0, OAL_MAX_SAMPLES * numBuffers * FRAME_SURROUND_FLOAT);
+	}
+	else
+	{
+		memset(sampleBuffer, 0, OAL_MAX_SAMPLES * numBuffers * FRAME_SURROUND_SHORT);
+	}
```
Comment (JosJuice) [inline Source/Core/AudioCommon/OpenALStream.cpp]: If you only have one line inside each block, you can write it like this:

    if (float32_capable)
        memset(sampleBuffer, 0, OAL_MAX_SAMPLES * numBuffers * FRAME_SURROUND_FLOAT);
    else
        memset(sampleBuffer, 0, OAL_MAX_SAMPLES * numBuffers * FRAME_SURROUND_SHORT);

This also applies to lines 150-153 and 174-181. Don't do it to lines 286-297.

File: Source/Core/AudioCommon/OpenALStream.cpp [hunk 4]
```diff
@@ -146,18 +146,39 @@ void OpenALStream::SoundLoop()
...
 	memset(realtimeBuffer, 0, OAL_MAX_SAMPLES * FRAME_STEREO_SHORT);
+
 	for (int i = 0; i < numBuffers; i++)
 	{
 		if (surround_capable)
```
Comment (lioncash) [inline Source/Core/AudioCommon/OpenALStream.cpp]: This if and its corresponding else needs to be braced.

File: Source/Core/AudioCommon/OpenALStream.cpp [hunk 5]
```diff
@@ -146,18 +146,39 @@ void OpenALStream::SoundLoop()
 	memset(uiBuffers, 0, numBuffers * sizeof(ALuint));
 	uiSource = 0;
 
+	// Checks if a X-Fi is being used. If it is, disable FLOAT32 support.
+	if (strstr(alGetString(AL_RENDERER), "X-Fi") != NULL)
+	{
+		float32_capable = false;
```
Comment (lioncash) [inline Source/Core/AudioCommon/OpenALStream.cpp]:
```cpp
float32_capable = (strstr(alGetString(AL_RENDERER), "X-Fi") == nullptr);
```

**Comment by phire** [comment]
The driver does what?

LGTM

**Comment by LAGonauta** [comment]
The X-Fi doesn't support floating point buffers, but the driver reports that the X-Fi has support for them. Sorry, I am not a native speaker and have some problems expressing myself.

**Pull request merged**